pkgs/development/haskell-modules/configuration-nix.nix

```diff
@@ -99,6 +99,9 @@ self: super: builtins.intersectAttrs super {
 
   niv = enableSeparateBinOutput super.niv;
 
+  # Ensure the necessary frameworks for Darwin.
+  OpenAL = addExtraLibrary super.OpenAL (pkgs.lib.optional pkgs.stdenv.isDarwin pkgs.darwin.apple_sdk.frameworks.OpenAL);
```

Comment:
This change should be similar to other changes already in the file:

```suggestion
  OpenAL = if pkgs.stdenv.isDarwin
    then addExtraLibrary super.OpenAL pkgs.darwin.apple_sdk.frameworks.OpenAL
    else super.OpenAL;
```

There is a small problem with the current code:

>  ```nix
> addExtraLibrary super.OpenAL (pkgs.lib.optional pkgs.stdenv.isDarwin pkgs.darwin.apple_sdk.frameworks.OpenAL);
> ```

Here, `addExtraLibrary` wants to take a derivation for its second argument.  (If you grep through the nixpkgs source code for `addExtraLibrary`, you can see that normally it takes things like `udev`, `libpcap`, etc.)

However, the output of `optional` is a list:

```console
$ nix repl '<nixpkgs>'
nix-repl> lib.optional true "foo"
[ "foo" ]
nix-repl> lib.optional false "foo"
[ ]
```

Now, in practice, I think that this still works out.  The derivations passed to `addExtraLibrary` eventually get passed to `buildInputs` for the Haskell package derivation.  Haskell package derivations eventually get built with `stdenv.mkDerivation`, which apparently knows how to deal with inputs in sublists:

```console
$ nix-shell -E 'with (import <nixpkgs> { }); runCommand "dummy" { buildInputs = [ gcc6 ]; } ""' --run 'gcc --version'
gcc (GCC) 6.5.0
$ nix-shell -E 'with (import <nixpkgs> { }); runCommand "dummy" { buildInputs = [ [ gcc6 ] ]; } ""' --run 'gcc --version'
gcc (GCC) 6.5.0
```

However, it is somewhat confusing to rely on this functionality when writing nix code, so it would be best to change the way it is currently written.

-----------------------------------------------------------------------------------------------------------------

As an aside, you could also do the following:

> ```nix
> addExtraLibraries super.OpenAL (pkgs.lib.optional pkgs.stdenv.isDarwin pkgs.darwin.apple_sdk.frameworks.OpenAL);
> ```

This works because `addExtraLibraries` expects a list of derivations as the second argument.

However, personally, I feel that just using an `if` statement is the clearest way to write it.  It is also used in other parts of the file already.

Reply:
Also, you may want to just rebase your commits into a single, clean commit.

Sorry for going back and forth so much on this change!

Reply:
No worries, I'm still new to nix so thanks for the patience. 

I'm a bit confused on what are the best practices regarding the overrides because both the `if-then` code and my original patch are already used.

Reply:
If possible, the "best practices" are to use the functions provided in `pkgs/development/haskell/lib.nix`.

However there are a lot of places not using the functions provided.  Ideally they would all be fixed where possible.